### PR TITLE
move dsn message integrity checking into dsn_message_parser

### PR DESCRIPTION
**Pull request by qinzuoyan**

### CODE REVIEW

**Comment by qinzuoyan** [comment]
@imzhenyu , please review it.

File: src/core/tools/common/dsn_message_parser.cpp [hunk 1]
```diff
@@ -113,7 +113,68 @@ namespace dsn
 
     int dsn_message_parser::prepare_on_send(message_ex* msg)
     {
-        return (int)msg->buffers.size();
+        auto& header = msg->header;
+        auto& buffers = msg->buffers;
+
+        if (task_spec::get(msg->local_rpc_code)->rpc_message_crc_required)
```
Comment (imzhenyu) [inline src/core/tools/common/dsn_message_parser.cpp:130]: This introduces performance issues as this function is called inside locks. Need to do this before linking the message into the sending queue.

**Comment by qinzuoyan** [comment]
@imzhenyu , now I have changed the interfaces of message_parser as:

```
        // prepare buffer before send.
        // this method should be called before get_buffer_count_on_send() and get_buffers_on_send()
        // to do some prepare operation.
        // may be invoked for many times if the message is resent.
        virtual void prepare_on_send(message_ex* msg) {}

        // get max buffer count needed by get_buffers_on_send().
        // may be invoked for many times if the message is resent.
        virtual int get_buffer_count_on_send(message_ex* msg) = 0;

        // get buffers from message to 'buffers'.
        // return buffer count used, which must be no more than the return value of get_buffer_count_on_send().
        // may be invoked for many times if the message is resent.
        virtual int get_buffers_on_send(message_ex* msg, /*out*/ send_buf* buffers) = 0;
```

Then method prepare_on_send() will be invoked out of lock.
And another thing to note is that the three methods may be invoked for more than one time, because the request message may be reused when resend.

File: src/core/tools/common/thrift_message_parser.cpp [hunk 2]
```diff
@@ -128,33 +116,38 @@ namespace dsn
...
-        //    |-----------response header--------|
+        // write thrift response header and thrift message begin
         binary_writer header_writer;
-        binary_writer_transport trans(header_writer);
-        boost::shared_ptr<binary_writer_transport> trans_ptr(&trans, [](binary_writer_transport*) {});
-        ::apache::thrift::protocol::TBinaryProtocol proto(trans_ptr);
-
-        //Total length, but we don't know the length, so firstly we put a placeholder
-        proto.writeI32(0);
+        binary_writer_transport header_trans(header_writer);
```
Comment (imzhenyu) [inline src/core/tools/common/thrift_message_parser.cpp:119]: can we move these logic into prepare_on_send as much as possible to reduce the lock conflict?

File: include/dsn/internal/rpc_message.h [hunk 3]
```diff
@@ -150,7 +150,8 @@ namespace dsn
     public:
         message_header         *header;
         std::vector<blob>      buffers; // header included for *send* message, 
-                                        // header not included for *recieved* 
+                                        // header not included for *recieved*
+        std::vector<blob>      buffers_for_parser;
```
Comment (imzhenyu) [inline include/dsn/internal/rpc_message.h]: This is the parser's problem, which introduces memory and performance overhead where normal rDSN msgs should not have. I prefer the original approach.

**Comment by imzhenyu** [comment]
The interface looks good to me. Btw, you may use `git reset --soft` and `git rebase` to remove unnecessary commits.

**Comment by qinzuoyan** [comment]
As your suggestion, I have moved most of logic from get_buffers_on_send() to prepare_on_send() for performance reason.

And the extra buffers generated by parser are appended to the back of message_ex::buffers.

I also considered the case that prepare_on_buffer() is invoked for more than once: the extra buffers at the end will be replaced on each preparing time.

**Comment by imzhenyu** [comment]
I see. We've used body_crc before to avoid computing the crc multiple times for the body part, can we borrow that for the preparation of those extra prefix/postfix things as well?

**Comment by qinzuoyan** [comment]
Not necessary, body crc is special because body will not changed even if resend, but header may be modified when resend (change remote address, timeout, etc), so it's simple and safe to compute header crc each time, and we can afford the overhead. So as the extra prefix/postfix things, the overhead of which is small.

**Comment by qinzuoyan** [comment]
I think it is an acceptable implementation now, and over optimization is not worthy. What do you think? @imzhenyu

**Comment by imzhenyu** [comment]
I mean, if we have multiple invocation of `prepare`, we may end up many (e.g., `2*N` where `N` is the sending count) blobs appended, which is unnecessary. You can argue that the overhead is not big, but I guess it is unnecessary and can be easily avoided?

**Comment by qinzuoyan** [comment]
Actually not, always only the last generated buffers will be left:

```
        // put header_bb and end_bb at the end
        buffers.resize(dsn_buf_count);
        buffers.emplace_back(std::move(header_bb));
        buffers.emplace_back(std::move(end_bb));
```

**Comment by imzhenyu** [comment]
Nice, then I'm ok, and the small optimization can be done later if necessary.

**Pull request merged**